### PR TITLE
Improved overflow support: Recursive directory watches

### PR DESCRIPTION
**Pull request by sungshik**
#### Note

To avoid diff noise, enable "Hide whitespace" when reviewing the code. (I pulled a few methods from an inner class to its outer class, so indentation changed.)

#### Overview

This PR:

  - Updates the existing `JDKRecursiveDirectoryWatch` to rely on the new mechanism to auto-handle overflows. These are the minimal changes required to make it work.
  - Adds a new `JDKFileTreeWatch` as an alternative recursive directory watch. I originally wrote this alternative as a proof-of-concept to better understand/design how auto-handling of overflows could be a standalone thing (i.e., also use it with file/non-recursive directory watches), and how the interaction with recursive directory watches would play out.

#### Task list

- [x] We should probably select one of the two watches and remove the other one to avoid unnecessary code growth
- [x] ~~If we choose `JDKRecursiveDirectoryWatch`, then it needs to be tidied up a bit (update comments etc.)~~

#### Similarities between the types of watches

Both types of recursive directory watches are containers for a collection of *internal* `JDKDirectoryWatch` objects (one for each subdirectory in scope), subject to the following principles:

- When a watch is started, recursively scan the directory and start an internal non-recursive `JDKDirectoryWatch` for each subdirectory.
- Between the recursive scan of the directory and the start of the internal watches, `CREATED` events for new subdirectories may have been missed. So, after starting the internal watches, send an `OVERFLOW` event  to the internal watch of the top-level directory. (It is forwarded to the other internal watches, as described below.) **Update:** This isn't needed for `JDKFileTreeWatch` (see the relevant comment in the code).
- When an event happens in an internal watch for subdirectory `x`, in addition to calling the event-handler:
    - `OVERFLOW`: Forward the event to the internal watches for subdirectories `y1`, `y2`, ..., of `x`.
    - `CREATED`, new subdirectory `y`: (1) Start an internal watch for `y`. (2) Between the creation of `y` and the start of the internal watch, events may have been missed. So, after starting the internal watch, send an `OVERFLOW` event to it.
    - `DELETED`, old subdirectory `y`: Stop the internal watch for `y`. 
    - `MODIFIED`: -

#### Differences

Using `JDKRecursiveDirectoryWatch`, the internal watches are stored "flat" inside a single `JDKRecursiveDirectoryWatch` object. Using `JDKFileTreeWatch`, the internal watches are stored "hierarchical" inside a tree of `JDKFileTreeWatch` objects (one internal watch per `JDKFileTreeWatch`).

For instance, suppose we have the following (sub)directories:
- `foo`
- `foo/bar1`
- `foo/bar2`
- `foo/bar3`
- `foo/bar3/baz1`
- `foo/bar3/baz2`

The structure of `JDKRecursiveDirectoryWatch` looks as follows:
```mermaid
graph TB
  subgraph Foo[JDKRecursiveDirectoryWatch]
    direction LR
    FooInner["JDKDirectoryWatch
    /foo"] ~~~
    Bar1Inner["JDKDirectoryWatch
    /foo/bar1"] ~~~
    Bar2Inner["JDKDirectoryWatch
    /foo/bar2"]
    Bar3Inner["JDKDirectoryWatch
    /foo/bar3"] ~~~
    Baz1Inner["JDKDirectoryWatch
    /foo/bar3/baz1"] ~~~
    Baz2Inner["JDKDirectoryWatch
    /foo/bar3/baz2"]
  end
```

The structure of `JDKFileTreeWatch` looks as follows:
```mermaid
graph TB
  subgraph Foo[JDKFileTreeWatch]
    FooInner["JDKDirectoryWatch
    /foo"]
  end

  subgraph Bar1[JDKFileTreeWatch]
    Bar1Inner["JDKDirectoryWatch
    /foo/bar1"]
  end

  subgraph Bar2[JDKFileTreeWatch]
    Bar2Inner["JDKDirectoryWatch
    /foo/bar2"]
  end

  subgraph Bar3[JDKFileTreeWatch]
    Bar3Inner["JDKDirectoryWatch
    /foo/bar3"]
  end

  subgraph Baz1[JDKFileTreeWatch]
    Baz1Inner["JDKDirectoryWatch
    /foo/bar3/baz1"]
  end

  subgraph Baz2[JDKFileTreeWatch]
    Baz2Inner["JDKDirectoryWatch
    /foo/bar3/baz2"]
  end

  Foo --> Bar1
  Foo --> Bar2
  Foo --> Bar3
  Bar3 --> Baz1
  Bar3 --> Baz2
```

### CODE REVIEW

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+
+        var internalEventHandler = eventHandler.andThen(new ChildWatchesUpdater());
+        this.internal = new JDKDirectoryWatch(path, exec, internalEventHandler) {
+
+            // Override to ensure that this watch relativizes events wrt
+            // `rootPath` (instead of `path`, as is the default behavior)
+            @Override
+            public WatchEvent relativize(WatchEvent event) {
+                return new WatchEvent(event.getKind(), rootPath,
+                    rootPath.relativize(event.calculateFullPath()));
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:72]: I think we should be able to calculate the relative path based on where we are in the tree? instead of doing a `calculateFullPath`?

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 2]
```diff
@@ -0,0 +1,221 @@
...
+     * child watch; (b) when a subdirectory creation happens, a new child watch
+     * is opened for that subdirectory; (c) when a subdirectory deletion
+     * happens, an existing child watch is closed for that subdirectory.
+     */
+    private class ChildWatchesUpdater implements BiConsumer<EventHandlingWatch, WatchEvent> {
+        @Override
+        public void accept(EventHandlingWatch watch, WatchEvent event) {
+            switch (event.getKind()) {
+                case OVERFLOW: acceptOverflow(); break;
+                case CREATED: acceptCreated(event.calculateFullPath()); break;
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:109]: For most places I think we should avoid calling `calculateFullPath` it should be more of a down-stream user function to call. It might be needed here, but since we have a current root path, we could just resolve it ourselves? which also avoids future relativize calls?

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+            }
+        }
+
+        private void acceptCreated(Path fullPath) {
+            if (Files.isDirectory(fullPath)) {
+                var childWatch = openChildWatch(fullPath);
+                // Events in the newly created directory might have been missed
+                // between its creation and setting up its watch. So, generate
+                // an `OVERFLOW` event for the watch.
+                reportOverflowTo(childWatch);
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:127]: With this strategy the semantic of this recursive watcher depend on the overflow handling mode. I would prefer to keep that separate. Not to mention we generate overflow events a lot more often. I get the proposal: well, this is like an overflow, so it's up to the user to deal with that?

Are we sure that the modification time tracker knows how to deal with an overflow right at (or after) its creation?

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+        var children = childWatches.keySet().iterator();
+        while (true) {
+            try {
+                // First, close all child watches
+                if (children.hasNext()) {
+                    closeChildWatch(children.next());
+                }
+                // Last, close the internal watch
+                else {
+                    internal.close();
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:190]: I think we should fist close our internal watch, to make sure we don't get new childeren registered on different threads, and then clear out chunks of the `children`, remove them from the map, and then loop untill map is empty. Maybe even set a `volatile boolean` field to mark this watchers as closed.

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+            children.forEach(this::openChildWatch);
+        } catch (IOException e) {
+            logger.error("File tree watch (for: {}) could not iterate over its children ({})", path, e);
+        }
+        // There's no need to report an overflow event, because `internal` was
+        // opened *before* the file system was accessed to fetch children. Thus,
+        // if a new directory is created while this method is running, then at
+        // least one of the following is true: (a) the new directory is already
+        // visible by the time the file system is accessed; (b) its `CREATED`
+        // event is handled later, which starts a new child watch if needed.
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:219]: yes, I hope this is true, sound like it should work

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+     * rules: (a) when an overflow happens, it's propagated to each existing
+     * child watch; (b) when a subdirectory creation happens, a new child watch
+     * is opened for that subdirectory; (c) when a subdirectory deletion
+     * happens, an existing child watch is closed for that subdirectory.
+     */
+    private class ChildWatchesUpdater implements BiConsumer<EventHandlingWatch, WatchEvent> {
+        @Override
+        public void accept(EventHandlingWatch watch, WatchEvent event) {
+            switch (event.getKind()) {
+                case OVERFLOW: acceptOverflow(); break;
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:108]: What I don't get is how we don't have to rescan the file system on overflow, to create child watchers for directories missed (and recursivly)

File: src/main/java/engineering/swat/watch/impl/jdk/JDKRecursiveDirectoryWatch.java [hunk 7]
```diff
@@ -85,27 +73,24 @@ private void handleCreate(WatchEvent ev) {
...
-            CompletableFuture
-                .completedFuture(fullPath)
-                .thenApplyAsync(this::registerForNewDirectory, exec)
-                .thenAcceptAsync(this::publishExtraEvents, exec)
-                .exceptionally(ex -> {
-                    logger.error("Could not locate new sub directories for: {}", ev.calculateFullPath(), ex);
-                    return null;
-                });
+            try {
+                if (Files.isDirectory(fullPath)) {
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKRecursiveDirectoryWatch.java:77]: This change now stops publication of events till after we've handled our internal bookkeeping.

The original code would schedule the book-keeping update on the background, to reduce the time it takes for an event to reach the library.

File: src/main/java/engineering/swat/watch/impl/jdk/JDKRecursiveDirectoryWatch.java [hunk 8]
```diff
@@ -85,27 +73,24 @@ private void handleCreate(WatchEvent ev) {
...
-        CompletableFuture
-            .completedFuture(ev.calculateFullPath())
-            .thenApplyAsync(this::syncAfterOverflow, exec)
-            .thenAcceptAsync(this::publishExtraEvents, exec)
-            .exceptionally(ex -> {
-                logger.error("Could not register new watch for: {} ({})", ev.calculateFullPath(), ex);
-                return null;
-            });
+        var fullPath = ev.calculateFullPath();
+        try (var children = Files.find(fullPath, 1, (p, attrs) -> p != fullPath && attrs.isDirectory())) {
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKRecursiveDirectoryWatch.java:89]: here the same, what if a user has configured to ignore overflow events? Does that mean we would then miss all nested directories in our watcher?

**Review by DavyLandman — CHANGES_REQUESTED** [review]
Nice! A lot of remove code. I like the idea to make it an actual tree, that will also reduce the pressure on the map. Although it will now cause multiple lookups in smaller maps. Almost makes you consider some kind of sortedmap/trie structure to quickly jump to the right child entry.

Some larger concerns I have with the current proposed changes:
1. It looks like the updating of the child directories (and especially missed nested child directories) depends on the overflow handling strategy. I do not think that is a wise dependency. Maybe we should add a test parameter to the torture test that cycles through all 3 enum values of the overflow handler
2. When implementing the original code my concern was to avoid iterating the filesystem more than needed, in profiling I noticed that had quite a perf hit, especially in larger directory hierarchies.
3. Similar point is with the scheduling of work on the Thread pool. My original approach has been to get the event that happened to the user as soon as we get them (as soon as we can), and do our internal bookkeeping in the background. For an IDE we don't want to have the event get in the back of the queue after our bookkeeping. I might be mistaken, but I have the feeling the current code is chaining a lot in `andThen` clauses without pushing them onto separate workjobs. If the design is that the original event gets pushed to the consumer first, and only after that is done do we handle the book-keeping, its a bit better, I mean, we might leave some cpu-cores idling, but at least we don't delay events.

File: src/main/java/engineering/swat/watch/impl/jdk/JDKRecursiveDirectoryWatch.java (hunk 7 above)
Comment (sungshik) [reply src/main/java/engineering/swat/watch/impl/jdk/JDKRecursiveDirectoryWatch.java]: Fixed. (Might require some more refactoring, but I need to tidy up this class a bit more anyway because of the changes in this PR.)

**Review by sungshik — COMMENTED** [review]
Thanks Davy! I addressed (and tentatively closed) all comments except [this one](https://github.com/SWAT-engineering/java-watch/pull/27/files/fa65b301252471d0f85b3e6f5ac7dfe00bb86ac9#r1991862012), which requires more thought...

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java (hunk 6 above)
Comment (sungshik) [reply src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java]: I made some progress to fix this issue for `JDKFileTreeWatch` in today's commits. Here's the [diff](https://github.com/SWAT-engineering/java-watch/pull/27/files/387e7c3520ecd0aeac94c9645196eb04d1e88d78..385db76368312dc771fccb64afb30e01f609438a). The fix itself is actually small and simple; most effort went into writing reliable tests. I've added them in a separate class (trying all three of `NONE`, `ALL`, and `DIRTY` for auto-handling of overflows). Anyway, these commits are ready for feedback.

Still working on a similar fix for `JDKRecursiveDirectoryWatch`. I'll finish that later this week.

File: src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java [hunk 9]
```diff
@@ -0,0 +1,236 @@
...
+            }
+            for (var file : createFiles.apply("file1.txt")) {
+                await("Creation of `" + file + "` should be observed (events: " + events + ")")
+                    .until(() -> !checkCreatedFiles || eventsContains.test(Kind.CREATED, file));
+            }
+
+            // Part 2: Artificially remove child watches. Changes in the root
+            // should still be observed by the watch, but changes in the
+            // descendants shouldn't.
+            getChildWatches(watch).remove(child1).close();
```
Comment (DavyLandman) [inline src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java:176]: I don't like tests that are so deeply integrated into the class they are testing. It makes us always having to update the tests for every small changes.

I prefer to have tests that verify the behavior is still what we expect and they only fail if we managed to break the externally observable behavior.

(especially as this is quite a large test case, I get that unease a bit more)

File: src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java [hunk 10]
```diff
@@ -0,0 +1,236 @@
...
+        }
+    }
+
+    @BeforeAll
+    static void setupEverything() {
+        Awaitility.setDefaultTimeout(TestHelper.NORMAL_WAIT);
+    }
+
+    @Test
+    void noRescannerPreservesIntegrityOfChildWatches() throws IOException {
```
Comment (DavyLandman) [inline src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java:83]: How about we devise of a torture test that also deletes and adds directories? and then we make it paramaterized where we pass in the `WatchScope` as an argument to the tests?

File: src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java (hunk 9 above)
Comment (sungshik) [reply src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java:176]: This isn't my preferred way of writing tests either of course 😎. It serves a purpose in this context, though.

The problem here is that we can't control whether or not the kernel misses events (and which ones) and produces an overflow instead. This makes it very tricky to predictably bring the test in the intended pre/post-states and affects the observations to be expected. I agree we can add a torture test and hope it triggers an overflow along the way, but: (a) this is unreliable/unpredictable; (b) when the test fails, the witness is rather indirect/coarse-grained, so it's not so clear what exactly goes wrong. Surely, the torture tests are useful in general, but for this particular case, a more fine-grained unit-like test was needed as well. If not to convince others, then at least to convince myself the *current* implementation works as intended -- and not by coincidence -- in at least the test case 😉.

The point that the test is sensitive to changes in the internals is valid, and so is the worry about (too) many false positives in the future. To (partially) mitigate this risk, already, I aimed to carefully document what exactly the test does, and what it expects. This could be construed as a partial specification of how the internals are intended to work today, so if something about that changes tomorrow, the comments should be helpful to identify the cause of the test failing. I expect this to be ok, so I'd keep the test in the suite for now, but I appreciate mileages may vary here (and/or change over time), so removing this test is fine by me, too.

Alternatively... I entertained the idea to "quickly" write our own mock `FileSystemProvider` to control exactly when overflow events happen (and which normal events are missed). This would allow us to better test all our overflow-related code (not just this), so there's some value to it. But it's also extra work.

Long story short:
 1. I can add a torture test (but it wouldn't serve exactly the same purpose)
 2. I can remove/disable the current test when the risk of future false positives is deemed too high
 3. I can implement a mock `FileSystemProvider` (technically the most flexible solution I think, short- and long-term, but also the most work)

Just let me know which combination of these three points you prefer 🙂.

File: src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java (hunk 10 above)
Comment (sungshik) [reply src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java:83]: See the previous comment (I hope)

File: src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java (hunk 9 above)
Comment (sungshik) [reply src/test/java/engineering/swat/watch/impl/JDKFileTreeWatchTests.java:176]: We're going for option 4: add a filtering mechanism to watches so we can simulate the kernel dropping events (and generating an overflow) using that filter. The test can then be rewritten in a reliable/predictable way without inspecting the internals.

**Review by sungshik — COMMENTED** [review]
Today's [series of commits](https://github.com/SWAT-engineering/java-watch/pull/27/files/385db76368312dc771fccb64afb30e01f609438a..76ec3800d62d2037cdbda3faf3bf1477c7c8b9a4) fix the following issues:
  - Primarily, it adds the filtering mechanism and revises the test (old one deleted, new one added)
  - Also, it fixes the closing issue we encountered (unrelated to filtering)
  - Also, it fixes a few small things related to relativization (also unrelated to filtering)

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 11]
```diff
@@ -0,0 +1,278 @@
...
+     * created subdirectories are opened, existing child watches for deleted
+     * subdirectories are closed, and the overflow is propagated to each child
+     * watch; (b) when a subdirectory creation happens, a new child watch is
+     * opened for that subdirectory; (c) when a subdirectory deletion happens,
+     * an existing child watch is closed for that subdirectory.
+     */
+    private class AsyncChildWatchesUpdater implements BiConsumer<EventHandlingWatch, WatchEvent> {
+        @Override
+        public void accept(EventHandlingWatch watch, WatchEvent event) {
+            exec.execute(() -> {
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:128]: this shouldn't happen if the parent has been closed.

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+        }
+
+        private void reportOverflowTo(JDKFileTreeWatch childWatch) {
+            var overflow = new WatchEvent(WatchEvent.Kind.OVERFLOW,
+                childWatch.rootPath, childWatch.relativePathParent);
+            childWatch.handleEvent(overflow);
+        }
+    }
+
+    private void openAndCloseChildWatches() {
```
Comment (DavyLandman) [inline src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:175]: nit: better name `syncWithFileSystem`? since now the name is saying what it does, but not why.

File: src/test/java/engineering/swat/watch/TestHelper.java [hunk 13]
```diff
@@ -48,9 +49,21 @@ else if (os.contains("win")) {
...
     }
 
+    public static void trySleep(Duration duration) {
+        trySleep(duration.toMillis());
+    }
+
+    public static void trySleep(long millis) {
+        try {
+            Thread.sleep(millis);
+        } catch (InterruptedException e) {
```
Comment (DavyLandman) [inline src/test/java/engineering/swat/watch/TestHelper.java:65]: [Never ignore/wrap an interrupted exception](https://www.baeldung.com/java-interrupted-exception). The best thing to do is to call:

```
Thread.currentThread().interrupt();
```

after we're done cleaning up.

**Review by DavyLandman — COMMENTED** [review]
A few small clarification questions.

File: src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java [hunk 14]
```diff
@@ -224,37 +252,19 @@ public void handleEvent(WatchEvent event) {
...
-                else {
-                    closeChildWatch(children.next());
-                }
-            } catch (IOException ex) {
-                logger.error("Could not close watch", ex);
-                firstFail = firstFail == null ? ex : firstFail;
-            } catch (Exception ex) {
-                logger.error("Could not close watch", ex);
-                firstFail = firstFail == null ? new IOException("Unexpected exception when closing", ex) : firstFail;
+        var firstFail = tryClose(internal);
```
[the remark this answers is not in the document]
Comment (sungshik) [reply src/main/java/engineering/swat/watch/impl/jdk/JDKFileTreeWatch.java:280]: Ha, that's a nice race, yes. As discussed, we'll continue to optimistically put new child watches in the map, but after that, we check if the parent watch wasn't closed in the meantime. I also added a comment to explain why this is needed.

**Review by DavyLandman — APPROVED** [review]
